### DISCUSSION TITLE
Put something like useShallow into the function not the caller

### DISCUSSION THREAD

**Discussion by Fabioni** [pmndrs/zustand#2430] [Q&A] [answered]
How can I make the memoization logic be part of the store function and not part of the component that calls the function.

```js
// store.js
const useStore = create((set, get) => ({
  bears: [{ id: 1, attitude: 'angry'}, { id: 2, attitude: 'playful'}]
  getFilteredBears: get().bears.filter(bear => bear.attitude === 'angry')  //! memoize that
}))

// component.js
function AngryBear() {
  const bears = useStore(state => state.getFilteredBears())
  return <h1>{bears.length} angry bears around here</h1>
}
```

right now I am using useShallow for that (`useStore(useShallow(state => state.getFilteredBears()))`), but I feel like logically the decision whether useShallow should be used or not belongs in the function itself and not in the caller.

**Comment by dai-shi** [comment]
You want to use some memoization library. `proxy-memoize` is one of mine.

```js
import { memoize } from 'proxy-memoize';

// store.js
const useStore = create((set, get) => ({
  bears: [{ id: 1, attitude: 'angry'}, { id: 2, attitude: 'playful'}]
}))
const getFilteredBears = memoize((state) => state.bears.filter((bear) => bear.attitude === 'angry'));

// component.js
function AngryBear() {
  const bears = useStore(getFilteredBears)
  return <h1>{bears.length} angry bears around here</h1>
}
```

**Reply by Fabioni** [reply]
From the website of proxy-memoize:
> The memoized function will re-evaluate the original function only if the used part of argument (object) is changed.

This means that the function **will** re-evaluate whenever `state.bears` changes. This is the correct behaviour, but not what I am looking for, since it will always trigger a re-render of the react component.

What `useStore(useShallow(state => state.getFilteredBears()))` is doing, is always re-evaluating the function but NOT triggering a re-rendering of the react component if the resulting objects of getFilteredBears are the same. This is the functionality I am looking for, and as far as I understand, nothing that proxy-memoize would help with.

**Reply by dai-shi** [reply, marked as answer]
I'd admit the docs aren't clear enough. https://redux.js.org/usage/deriving-data-selectors#proxy-memoize might describe it better.

That said, yeah, `proxy-memoize` is just one example. If you only care about the result being shallow equal, it's much easier.
```js
const memoize = (fn) => {
  let prev;
  return (...args) => {
    const result = fn(...args);
    return shallow(prev, result) ? prev : (prev = result);
  };
};
```
